core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala

```diff
@@ -289,13 +289,10 @@ class BrokerMetadataPublisher(
     try {
       // Start log manager, which will perform (potentially lengthy)
       // recovery-from-unclean-shutdown if required.
-      logManager.startup(metadataCache.getAllTopics())
-
-      // Delete partition directories which we're not supposed to have. We have
-      // to do this before starting ReplicaManager, so that the stray replicas
-      // don't block creation of new ones with different IDs but the same names.
-      // See KAFKA-14616 for details.
-      logManager.deleteStrayKRaftReplicas(brokerId, newImage.topics())
+      logManager.startup(
+        metadataCache.getAllTopics(),
+        shouldBeStrayKraftLog = log => LogManager.isStrayKraftReplica(brokerId, newImage.topics(), log)
```

Comment:
A suggestion, due to the following concerns:

* LogManager shouldn't handle metadata records, these types shouldn't depend on each other. The metadata records should be handled here instead.
* The argument name looks a bit strange, namely the 'should' and 'kraft' parts.

```suggestion
        isStray = (topicId, partition) => Option(newImage.topics().getPartition(topicId.getOrElse{
          throw new RuntimeException(s"Partition $partition does not have a topic ID, " +
            "which is not allowed when running in KRaft mode.")
        }, partition.partition())).exists(_.replicas.contains(brokerId))
```

Perhaps LogManager should declare a type for this argument since it's propagated down the call stack several levels?

Reply:
I updated this with the suggested comment.